Recovery auth codes

rh-iatanaso opened the pull request:
<!---
Please read https://github.com/keycloak/keycloak/blob/main/CONTRIBUTING.md and follow these guidelines when contributing to Keycloak
-->
 Initial implementation for the Recovery codes base on https://github.com/keycloak/keycloak/discussions/8518

Closes #9540

stianst commented on server-spi-private/src/main/java/org/keycloak/forms/login/LoginFormsPages.java:
```diff
@@ -27,6 +27,7 @@
     LOGIN_OAUTH2_DEVICE_VERIFY_USER_CODE, UPDATE_USER_PROFILE, IDP_REVIEW_USER_PROFILE,
-    FRONTCHANNEL_LOGOUT;
+    FRONTCHANNEL_LOGOUT,
+    LOGIN_RECOVERY_AUTHN_CODES_INPUT, LOGIN_RECOVERY_AUTHN_CODES_CONFIG;
```
Should rather just use the update credentials event than to introduce a dedicate even for recovery codes

rh-iatanaso replied:
@stianst I think this Enum drives the Freemaker templates and not the events. I don't think we created custom event for recovery codes

rh-iatanaso committed:
Initial implementation for the Recovery codes base on https://github.com/keycloak/keycloak/discussions/8518

Recovery auth codes modify after review (#17)

* CIAM-2154: Make required action default. (#7)

* CIAM-2162: shorten help text message. (#9)

* CIAM-2197: Remove direct grant. (#11)

* CIAM-2197: Missed removing direct grant stuff from services. (#13)

* cleared out the label for remaining recovery codes (#12)

Co-authored-by: zwitter <zwitter@redhat.com>

* CIAM-2154: Change recovery code required action default priority. (#16)

* CIAM-2155: set updateAction to make codes updatable. (#8)

* CIAM-2153: Tie required action to session instead of user. (#10)

* Ciam 2165 make generic metadata (#14)

* CIAM-2165: work in progress.

* CIAM-2165: add missing import.

* CIAM-2165: minor cleanup.

* CIAM-2165: first attempt at getting signing page to use new data structure.

* CIAM-2165: Add missing java file.

* CIAM-2165: default to null instead of null string.

* CIAM-2165: fix messages.

* CIAM-2165: Clean up exception handling.

* CIAM-2165: Stop adding metadata to credentialRepresentation.

* CIAM-2165: Fix tabbing.

* CIAM-2165: Fix tabbing again.

* CIAM-2161: remove unused messages. (#15)

* CIAM-2161: remove unused messages.

* CIAM-2161: Remove old messages from previous version of recovery codes.

* CIAM-2161: Remove old messages from previous version of backup codes.

* CIAM-2161: Remove backupcodesIntroMessage from properties.

* Ciam 2165 make generic metadata (#18)

* CIAM-2165: work in progress.

* CIAM-2165: minor cleanup.

* CIAM-2165: first attempt at getting signing page to use new data structure.

* CIAM-2165: default to null instead of null string.

* CIAM-2165: fix messages.

* CIAM-2165: Clean up exception handling.

* CIAM-2165: Stop adding metadata to credentialRepresentation.

* CIAM-2165: Fix tabbing.

* CIAM-2165: Fix tabbing again.

* CIAM-2165: Fix failing tests.

* CIAM-2165: Improve log message.

* CIMA-2165: Remove duplicate method.

* CIAM-2165: Remove unused interface.

* CIAM-2165: skip secretdata readValue if it's null.

Co-authored-by: Zachary Witter <torquekma@gmail.com>
Co-authored-by: zwitter <zwitter@redhat.com>

CIAM-2165: Fix npe that could happen. (#19)

CIAM-2295: Make suggested changes from reviewers. (#21)

* CIAM-2295: Make suggest changes from Thomas.

* CIAM-2295: Make threshold configurable via password policy.

* CIAM-2295: Make other changes requested.

* CIAM-2295: More cleanup.

* CIAM-2295: Clarify the message for recovery codes used/total.

* CIAM-2295: Remove comment.

* CIAM-2295: Get feature flag working.

* CIAM-2295: add feature flag for recovery codes.

* CIAM-2295: restore whitespace.

rh-iatanaso commented:
@mposolda @stianst  Can I get approval for running the workflows . Also most of the feedback I can incorprate from the discussion is in the PR

rh-iatanaso commented:
@stianst @mposolda Can I get anothe approval for the work flow , we had issue which we addressed.

rh-iatanaso commented:
@stianst @mposolda Can I get another  workflow approval for this PR ?

mposolda commented on services/src/main/java/org/keycloak/authentication/authenticators/browser/RecoveryAuthnCodesFormAuthenticator.java:
```diff
@@ -0,0 +1,153 @@
+
+    @Override
+    public void setRequiredActions(KeycloakSession session, RealmModel realm, UserModel user) {
+        user.addRequiredAction(UserModel.RequiredAction.CONFIGURE_RECOVERY_AUTHN_CODES.name());
```
I think it will be better if required action is added to the authentication session rather than to the user. See `OTPFormAuthenticator` or `WebAuthnAuthenticator` classes for the inspiration. Or is there any reason why for backup codes it needs to be done differently than for OTP/WebAuthn?

stianst replied:
It should be added to auth-session right, otherwise we're updating the user session when that's not needed, and also it would linger in a subsequent login, which isn't correct.

stelewis-redhat replied:
What priority should I give the required action for recovery codes? Should it be very low or should it be near where the Configure OTP action is in priority?

mposolda replied:
Hmm.... I am not sure what priority you have in mind? Could you please elaborate? This note was about that instead of calling `user.addRequiredAction(UserModel.RequiredAction.CONFIGURE_RECOVERY_AUTHN_CODES.name())` you can call something like `authenticationSession.addRequiredAction(UserModel.RequiredAction.CONFIGURE_RECOVERY_AUTHN_CODES.name())` to make sure that adding required action is scoped only to this authentication and not to user himself. There is no any priority involved unless I am missing something?

stelewis-redhat replied:
I must have commented on the wrong note. I understand what to do about removing user.addRequiredAction.

My question was about the note above that about making the required action default. There is a priority that's a number. Right now, when you register a required action it gives it the lowest priority.

See the MR at https://github.com/rh-iatanaso/keycloak/pull/7/files. In this MR I've set it to 5 but that's almost certainly wrong.

mposolda commented on services/src/main/java/org/keycloak/authentication/authenticators/browser/RecoveryAuthnCodesFormAuthenticator.java:
```diff
@@ -0,0 +1,153 @@
+import org.keycloak.events.Details;
+import org.keycloak.events.Errors;
+import org.keycloak.forms.login.LoginFormsProvider;
+import org.keycloak.models.*;
```
Minor: Star imports should be avoided in Keycloak project (It is the rule, which we try to follow in Keycloak, even if it is probably not officially documented anywhere...)

mposolda commented on testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/pages/AuthenticationMethodSetupPage.java:
```diff
@@ -0,0 +1,54 @@
+ *
+ * @author <a href="mailto:vnukala@redhat.com">Venkata Nukala</a>
+ */
+public class AuthenticationMethodSetupPage extends LanguageComboboxAwarePage {
```
This class is used in RecoveryAuthnCodesAuthenticatorTest, but this test does not use it for anything. So since this class is effectively not used anywhere, it can be possible removed from this PR?

mposolda commented on testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/pages/LandingPage.java:
```diff
@@ -0,0 +1,50 @@
+ *
+ * @author <a href="mailto:vnukala@redhat.com">Venkata Nukala</a>
+ */
+public class LandingPage extends LanguageComboboxAwarePage {
```
This class is used in RecoveryAuthnCodesAuthenticatorTest, but this test does not use it for anything. So since this class is effectively not used anywhere, it can be possible removed from this PR?

mposolda commented on testsuite/utils/src/main/java/org/keycloak/testsuite/TotpGenerator.java:
```diff
@@ -34,7 +34,7 @@ public static void main(String[] args) throws IOException {
         Timer timer = new Timer();
         TotpTask task = null;
-
+        String secret = null;
```
Minor: This line seems to be redundant and can be removed. Class `TotpGenerator` likely doesn't need to be changed in this PR

mposolda reviewed (changes requested):
@rh-iatanaso I think this is great work. Congratulations to you and to the team!

I've added few comments inline. Most of them are minor things.

Besides that, I have one concern: Do we want `Recovery Authentication Codes` required action to be registered by default in the realms? I think that it will be good similarly like `Configure OTP` action is registered by default. I know that WebAuthn related actions are not registered by default, but that is related to the fact that WebAuthn was not supported. However the "Recovery codes" are supported from day 1 and hence required action might be good to be registered by default from day 1?

stianst commented on services/src/main/java/org/keycloak/authentication/authenticators/browser/RecoveryAuthnCodesFormAuthenticator.java:
```diff
@@ -0,0 +1,153 @@
+        return result;
+    }
+
+    protected boolean isDisabledByBruteForce(AuthenticationFlowContext authnFlowContext, UserModel authenticatedUser) {
+        boolean bolResult = false;
+        String bruteForceError;
+        Response challengeResponse;
+        bruteForceError = AuthenticatorUtils.getDisabledByBruteForceEventError(authnFlowContext.getProtector(),
+                authnFlowContext.getSession(), authnFlowContext.getRealm(), authenticatedUser);
+        if (bruteForceError != null) {
+            authnFlowContext.getEvent().user(authenticatedUser);
+            authnFlowContext.getEvent().error(bruteForceError);
+            challengeResponse = createLoginForm(authnFlowContext, false, Messages.INVALID_USER, FIELD_USERNAME);
+            authnFlowContext.forceChallenge(challengeResponse);
+            bolResult = true;
+        }
+        return bolResult;
+    }
```
@mposolda does an authenticator have to deal with brute-force itself? I would have thought that would be built-into the flow somehow?

mposolda replied:
I was thinking the same, but it is not :/ The individual authenticators currently checks by itself (For example `AbstractUsernameFormAuthenticator.isDisabledByBruteForce` ).

I don't know what was the motivation to do it this way? I guess it can be related to the fact that you still want to be able to authenticate user for example by `CookieAuthenticator` (or some others) even if he is temporarily disabled.

So for example flow like:
- Valid user authenticates and creates session1
- Attacker tries brute forcing in different laptop and this cause user to be temporarily disabled
- Valid user is currently still able to SSO re-authenticate in session1 even if he is disabled by brute force.
 
If we check brute force at the flow level for every authenticator, it would mean that SSO re-authentication with cookie may not work and hence attacker has "power" to break the valid SSO sessions of existing users. Which is also not ideal...

stelewis-redhat replied:
Is there consensus on whether we should leave it as is, or remove the brute-force code?

stianst replied:
We should leave it as is since there's no "global" check for it. @mposolda I think it may be worth re-visiting in the future, so it's not individual authenticators that control brute-force, but I do agree that the "cookie" is a special case though

stelewis-redhat replied:
@stianst "Neither admin or account console shows how many codes are left" the account console shows how many codes are left. Where would we show it on the admin console? Is that something we can revisit later?

stianst commented on services/src/main/java/org/keycloak/authentication/authenticators/directgrant/ValidateRecoveryAuthnCode.java:
```diff
@@ -0,0 +1,142 @@
+import java.util.LinkedList;
+import java.util.List;
+
+public class ValidateRecoveryAuthnCode extends AbstractDirectGrantAuthenticator
```
I don't think we should support recover codes at all with direct-grant, so I would actually suggest removing this one completely?

mposolda replied:
I was wondering about the same yesterday...

However I was thinking about the case when 2-factor is required for browser login and at the same time, you have "Direct Grant" enabled for some clients. Isn't it strange that browser login will require 2-factor, but direct-grant allows to login with simple username/password? Hence it makes sense to me to have authenticator for direct-grant as well (similarly like we have for OTP).

rh-iatanaso replied:
This is exactly what we were thinking that it would just be strange not to be able to use recovery codes since you can use it for OTP

stianst commented on themes/src/main/resources/theme/base/account/messages/messages_en.properties:
```diff
@@ -337,15 +337,15 @@ sendVerficationCode=Send Verification Code
-authenticatorBackupCodesSetupTitle=Backup Codes Setup
-backupcodesIntroMessage=If you lose access to your phone, you can still log into your account through backup codes. Keep them somewhere safe and accessible.
+authenticatorBackupCodesSetupTitle=Recovery Authentication Codes Setup
+backupcodesIntroMessage=If you lose access to your phone, you can still log into your account through recovery codes. Keep them somewhere safe and accessible.
```
I can't see where quite a few of these message keys are used. Like backupcodesIntroMessage and backupCodesTips?

rh-iatanaso replied:
Those where old messages from older implementation, I can remove the messages as part of this PR

stianst replied:
Either remove them not or not update them. I suspect these where used by OTP a long time ago.

stianst commented on themes/src/main/resources-community/theme/base/account/messages/messages_pt_BR.properties:
```diff
@@ -293,9 +293,9 @@ authenticatorSMSMessage=A aplica\u00e7\u00e3o ir\u00e1 enviar o c\u00f3digo de v
-authenticatorBackupCodesTitle=C\u00f3digos de Emerg\u00eancia
-authenticatorBackupCodesMessage=Veja seus c\u00f3digos de emerg\u00eancia de 8 d\u00edgitos
-authenticatorBackupCodesFinishSetUpMessage=12 c\u00f3digos de emerg\u00eancia foram gerados. Cada um pode ser utilizado apenas uma vez.
+authenticatorRecoveryAuthnCodesTitle=C\u00f3digos de Emerg\u00eancia
+authenticatorRecoveryAuthnCodesMessage=Veja seus c\u00f3digos de emerg\u00eancia de 12 d\u00edgitos
+authenticatorRecoveryAuthnCodesFinishSetUpMessage=12 c\u00f3digos de emerg\u00eancia foram gerados. Cada um pode ser utilizado apenas uma vez.
```
I assume all the translations are just google translate or something? We don't usually add translations, but require the community to contribute those so they can be reviewed properly

stianst replied:
Hmm.. had no clue that we had ever had anything around backup codes. You sure these are not related to the old way of setting up OTP?

stianst commented on themes/src/main/resources/theme/keycloak.v2/account/messages/messages_en.properties:
```diff
@@ -89,6 +89,9 @@ password-help-text=Log in by entering your password.
 otp-help-text=Enter a verification code from authenticator application.
+recovery-authn-code=My Recovery Authentication Codes
+recovery-authn-codes-display-name=Recovery Authentication Codes
+recovery-authn-codes-help-text=Recovery Authentication Codes can be used as a second factor or to recover access to your account in the event of a lost second factor.
```
This should be made a bit shorter

rh-iatanaso replied:
This sounds better -> "These codes can be used to regain your access in case your other 2FA means are not available"

stianst replied:
+1

stianst reviewed (changes requested):
* Documentation and release notes need to be updated for this with a separate PR to keycloak-documentation
* I would remove the label (My generated recovery authentication codes0) as it has no practical use for a user as there can only be one set of recovery codes
* User can't update the codes, but have to remove and re-create. It should be made updatable and allow generating new codes when the user clicks the update button
* Neither admin or account console shows how many codes are left

mposolda commented on server-spi/src/main/java/org/keycloak/models/credential/dto/RecoveryAuthnCodesCredentialData.java:
```diff
@@ -0,0 +1,48 @@
+                                            @JsonProperty("total") int totalCodes) {
+        this.hashIterations = hashIterations;
+        this.algorithm = algorithm;
+        this.remainingCodes = remainingCodes;
```
It looks there is missing line:
```
this.totalCodes = totalCodes;
```
This causes that admin console always shows 0 total codes, which is incorrect.

mposolda commented:
@stianst @rh-iatanaso Added some inline comments:

> * Documentation and release notes need to be updated for this with a separate PR to keycloak-documentation
> 
>     * I would remove the label (My generated recovery authentication codes0) as it has no practical use for a user as there can only be one set of recovery codes

+1 . However currently account console is hardcoded to display  credential label. I think that if we want to display remaining count of codes instead of label, it will require some changes at the SPI level.

I think that CredentialTypeMetadata class can have new field like `accountConsoleLabel`, which will by default fallback to credential user label. For Backup codes credential, it can instead show something like `Remaining codes count: 10` text (localized to proper user locale).

This will require few changes though, so question is, if we want to do it as follow-up or still have it in this PR?

> 
>     * User can't update the codes, but have to remove and re-create. It should be made updatable and allow generating new codes when the user clicks the update button

+1 . I think current SPI allows this if you use `CredentialTypeMetadata.updateAction` for recovery code credential provider?

> 
>     * Neither admin or account console shows how many codes are left

Admin console displays it when you click on the credential and click "Show data", it shows remaining codes count. Now I figured there is bug with "total codes", which displays incorrect count (I've commented in the inline comment for this).

For account console, I commented above.

mposolda commented on themes/src/main/resources/theme/keycloak.v2/account/src/app/content/signingin-page/SigningInPage.tsx:
```diff
@@ -249,7 +259,32 @@ class SigningInPage extends React.Component<SigningInPageProps, SigningInPageSta
+                            <div className="pf-c-alert__icon">
+                                <i className="pficon-warning-triangle-o" aria-hidden="true"></i>
+                            </div>
+                            <h4 className="pf-c-alert__title">
```
It will be ideal if we can avoid the account-console code specific for "recovery-code" credential type.

The account console is supposed to be used with custom credential type implementations provided by administrator. Having the code specific for each credential does not scale IMO. Will be good if we can handle all of this with credential metadata.

rh-iatanaso replied:
I can move the userLabel generation inside the code. I am not sure how to call the alert in agnostic way.

mposolda replied:
@stelewis-redhat Yes, the second thing you mentioned. I am thinking about CredentialMetadata as separate object, which will contain "CredentialModel" as one of the fields. See my comment above.

The main motivation is that `SigningInPage.tsx` doesn't need to contain any if/else blocks or other hacks specific to `Recovery-code` credential. As it will be ideal that customers, who potentially implement their own credentials/authenticators don't need to change UI classes, but SPI is good enough so they can add their own info/warning labels shown alongside their credential just by implementing this method on `CredentialProvider` interface.

stelewis-redhat replied:
OK. I can switch that around. I have a working solution for the first, but it shouldn't be too hard to swap that structure.

stelewis-redhat replied:
@mposolda Is it possible to revisit making CredentialMetadata contain CredentialModel, after we do the account console redesign? We're a bit worried about the merge/rebasing issues with these two initiatives going on at once.

Here's what I currently have working, for reference: https://github.com/rh-iatanaso/keycloak/pull/14 not final, work in progress.

mposolda replied:
I vote to avoid doing changes in `CredentialModel` and `CredentialRepresentation` classes and rather do it the other way around.
The problem with changing `CredentialModel` is, that the class is part of public SPI and part of the model. This mean that if we change anything in it, it will become part of "public SPI" and it is a pain to remove/change it later. So ideal would be if the initial commit already is done in a way that `CredentialModel` does not need to be changed.

stelewis-redhat replied:
OK. Thank you @mposolda

stelewis-redhat commented:
> 

For the required action as default, what priority should it have? Same as OTP? Or something else?

stelewis-redhat commented:
@stianst @mposolda I've merged in the changes for the requested updates.

stelewis-redhat commented on services/src/main/java/org/keycloak/services/resources/account/AccountCredentialResource.java:
```diff
@@ -180,29 +182,41 @@ public CredentialTypeMetadata getMetadata() {
+                List<CredentialMetadata> credentialMetadataList = modelsOfType.stream()
+                        .map(m -> {
+                            return credentialProvider.getCredentialMetadata(
+                                    credentialProvider.getCredentialFromModel(m), metadata
```
I don't love having to go get the data again here, but I think that's necessary, at least with what I understand right now.

mposolda replied:
I think this is OK to me, at least for now.

stelewis-redhat commented on server-spi/src/main/java/org/keycloak/models/credential/OTPCredentialModel.java:
```diff
@@ -59,7 +59,10 @@ public static OTPCredentialModel createFromPolicy(RealmModel realm, String secre
             OTPCredentialData credentialData = JsonSerialization.readValue(credentialModel.getCredentialData(), OTPCredentialData.class);
-            OTPSecretData secretData = JsonSerialization.readValue(credentialModel.getSecretData(), OTPSecretData.class);
+            OTPSecretData secretData = null;
+            if (credentialModel.getSecretData() != null) {
```
Since the credentialModel may have the secret removed for security reasons, I'm having to add this guard clause.

stelewis-redhat replied:
> > @stelewis-redhat let's not expand the scope of this PR to make everything configurable. I question that is needed in general, and it will also grow the scope of this PR a lot. So that should be a follow-up if not. For now let's focus on getting this ready to be merged.
> 
> Sorry, I didn't want to blow up the scope of this PR. However, I think being able to configure essential settings is an integral part of such a feature. But I agree, the feature is usable even without support for explicit configuration. However this should IMHO be added soon for completeness.
> 
> Some general guidance on feature completeness might help in the future? If we are striving to get something working initially, but it's not production grade then we should put it behind a preview flag. When something is "production grade" has to be defined.

OK I think I have everything done, except for turning on the feature flag.

stelewis-redhat commented on server-spi/src/main/java/org/keycloak/models/credential/RecoveryAuthnCodesCredentialModel.java:
```diff
@@ -0,0 +1,106 @@
+        try {
+            credentialData = JsonSerialization.readValue(credentialModel.getCredentialData(),
+                    RecoveryAuthnCodesCredentialData.class);
+            if (credentialModel.getSecretData() != null) {
```
Since the credentialModel may have the secret removed for security reasons, I'm having to add this guard clause.

thomasdarimont replied:
This pattern seems to be used multiple times - perhaps a common helper method could help to remove some code duplication.

stelewis-redhat replied:
I believe I can remove this if-statement completely.

stelewis-redhat commented:
@thomasdarimont Do you have any further comments? Also, do you have any time to do a proper review of the code? Thank you.

thomasdarimont commented on server-spi/src/main/java/org/keycloak/credential/CredentialMetadata.java:
```diff
@@ -0,0 +1,63 @@
+        this.infoMessage = message;
+    }
+
+    class LocalizedMessage {
```
If this type is exposed via Public API, e.g. #getInfoMessage(), then the type should be `public static`

stelewis-redhat replied:
OK.

thomasdarimont commented on server-spi/src/main/java/org/keycloak/credential/CredentialMetadata.java:
```diff
@@ -0,0 +1,63 @@
+    }
+
+    class LocalizedMessage {
+        String key;
```
fields can be private final

stelewis-redhat replied:
OK.

thomasdarimont commented on server-spi/src/main/java/org/keycloak/models/UserCredentialModel.java:
```diff
@@ -126,6 +127,14 @@ public static UserCredentialModel generateSecret() {
     }
 
+    public static UserCredentialModel buildFromBackupAuthnCode(String backupAuthnCodeInput) {
+        return new UserCredentialModel("", RecoveryAuthnCodesCredentialModel.TYPE, backupAuthnCodeInput);
```
you could call `buildFromBackupAuthnCode("", backupAuthnCodeInput)` here

stelewis-redhat replied:
OK.

thomasdarimont commented on server-spi/src/main/java/org/keycloak/models/credential/RecoveryAuthnCodesCredentialModel.java:
```diff
@@ -0,0 +1,106 @@
+
+    private final RecoveryAuthnCodesCredentialData credentialData;
+    private final RecoveryAuthnCodesSecretData secretData;
+    private final CredentialMetadata credentialMetadata = null;
```
This seems to be unused, remove?

stelewis-redhat replied:
OK.

thomasdarimont commented on services/src/main/java/org/keycloak/authentication/authenticators/browser/RecoveryAuthnCodesFormAuthenticator.java:
```diff
@@ -0,0 +1,159 @@
+        if (ObjectUtil.isBlank(recoveryAuthnCodeUserInput)) {
+            authnFlowContext.forceChallenge(createLoginForm(authnFlowContext, true,
+                    RecoveryAuthnCodesUtils.RECOVERY_AUTHN_CODES_INPUT_DEFAULT_ERROR_MESSAGE,
+                    RecoveryAuthnCodesUtils.FIELD_RECOVERY_CODE_IN_BROWSER_FLOW));
```
early exit here could help to reduce nesting and make the code easier to follow.

stelewis-redhat replied:
OK.

thomasdarimont commented on services/src/main/java/org/keycloak/authentication/authenticators/browser/RecoveryAuthnCodesFormAuthenticator.java:
```diff
@@ -0,0 +1,159 @@
+        Response challengeResponse;
+        bruteForceError = AuthenticatorUtils.getDisabledByBruteForceEventError(authnFlowContext.getProtector(),
+                authnFlowContext.getSession(), authnFlowContext.getRealm(), authenticatedUser);
+        if (bruteForceError != null) {
```
How about using early exit here as well?

thomasdarimont replied:
You could also use the occasion to introduce a new overload like `AuthenticatorUtils.getDisabledByBruteForceEventError(AuthenticationFlowContext authnFlowContext, UserModel authenticatedUser)` to . This could help to streamline the API usage of a lot of other authenticator implementations.

stelewis-redhat replied:
OK.

thomasdarimont commented on services/src/main/java/org/keycloak/credential/RecoveryAuthnCodesCredentialProvider.java:
```diff
@@ -0,0 +1,118 @@
+
+    @Override
+    public boolean isConfiguredFor(RealmModel realm, UserModel user, String credentialType) {
+        return session.userCredentialManager().getStoredCredentialsByTypeStream(realm, user, credentialType).findAny()
```
```suggestion
        return session.userCredentialManager().getStoredCredentialsByTypeStream(realm, user, credentialType).anyMatch(Objects::nonNull);
```

thomasdarimont commented on services/src/main/java/org/keycloak/credential/RecoveryAuthnCodesCredentialProvider.java:
```diff
@@ -0,0 +1,118 @@
+
+    @Override
+    public boolean isConfiguredFor(RealmModel realm, UserModel user, String credentialType) {
+        return session.userCredentialManager().getStoredCredentialsByTypeStream(realm, user, credentialType).findAny()
```
```suggestion
               return session.userCredentialManager().getStoredCredentialsByTypeStream(realm, user, credentialType).anyMatch(Objects::nonNull);
```

stelewis-redhat replied:
OK.

thomasdarimont commented on server-spi/src/main/java/org/keycloak/models/credential/RecoveryAuthnCodesCredentialModel.java:
```diff
@@ -0,0 +1,106 @@
+                            RecoveryAuthnCodesUtils.hashRawCode(originalGeneratedCodes.get(i))))
+                    .collect(Collectors.toList());
+            secretData = new RecoveryAuthnCodesSecretData(recoveryCodes);
+            credentialData = new RecoveryAuthnCodesCredentialData(RecoveryAuthnCodesUtils.NUM_HASH_ITERATIONS,
```
Some environments might have stronger requirements then this defaults.

Could we make the number of iterations and the hash mechanism configurable, e.g. via password policies?

Alternatively extract the call to 
```
RecoveryAuthnCodesCredentialModel.createFromValues(generatedCodes,
                generatedAtTime, generatedUserLabel);
``` in `RecoveryAuthnCodesAction` into a protected method to allow users to apply a custom mechanism.

stelewis-redhat replied:
OK.

stelewis-redhat commented on themes/src/main/resources/theme/base/login/login-recovery-authn-code-config.ftl:
```diff
@@ -0,0 +1,184 @@
+    <form action="${url.loginAction}" class="${properties.kcFormClass!}" id="kc-recovery-codes-settings-form" method="post">
+        <input type="hidden" name="generatedRecoveryAuthnCodes" value="${recoveryAuthnCodesConfigBean.generatedRecoveryAuthnCodesAsString}" />
+        <input type="hidden" name="generatedAt" value="${recoveryAuthnCodesConfigBean.generatedAt?c}" />
+        <input type="hidden" name="userLabel" value=" " />
```
Not sure offhand. I didn't add that myself. Will try to figure it out. I see some Java code that pulls those 3 hidden values.

thomasdarimont commented:
The currently implementation shows the number of codes used instead of the number of remaining codes.
![image](https://user-images.githubusercontent.com/314690/156231878-55e8239d-fdc4-48a1-9d8d-a5f0ccc38423.png)
Is it also possible to get the information how many codes remain? In some customer projects I implemented similar functionality and there users see a warning if the number of remaining codes is lower than a threshold (e.g. 2)

stelewis-redhat commented:
> The currently implementation shows the number of codes used instead of the number of remaining codes. ![image](https://user-images.githubusercontent.com/314690/156231878-55e8239d-fdc4-48a1-9d8d-a5f0ccc38423.png) Is it also possible to get the information how many codes remain? In some customer projects I implemented similar functionality and there users see a warning if the number of remaining codes is lower than a threshold (e.g. 2)

Yes, there is a warning message once it drops below 4.

thomasdarimont commented:
> Yes, there is a warning message once it drops below 4.
Is this already configurable? I didn't see that in the code :)

thomasdarimont commented on server-spi/src/main/java/org/keycloak/models/credential/RecoveryAuthnCodesCredentialModel.java:
```diff
@@ -0,0 +1,106 @@
+    }
+
+    public RecoveryAuthnCodeRepresentation getNextRecoveryAuthnCode() {
+        return this.secretData.getCodes().get(0);
```
If getCodes() retruns an empty list, ArrayIndexOutOfBoundsException will be thrown. We could wrap the return in an `Optional` or simply return `null`.

Is the idea to require users always call `allCodesUsed()` before caling `getNextRecoveryAuthnCode()`?

stelewis-redhat replied:
I'll use an optional. Good idea.

stelewis-redhat commented:
> > Yes, there is a warning message once it drops below 4.
> > Is this already configurable? I didn't see that in the code :)

No. What's the best mechanism to make it configurable? Use a properties file? Or something else?

thomasdarimont commented:
> No. What's the best mechanism to make it configurable? Use a properties file? Or something else?

I think one way could be to "abuse" a password policy here. Currently users can configure the iteration count and default password hashing mechanism via policies:

![image](https://user-images.githubusercontent.com/314690/156236839-cb990945-0e1c-4420-97f1-6045b432b44a.png)

stelewis-redhat commented:
> > No. What's the best mechanism to make it configurable? Use a properties file? Or something else?
> 
> I think one way could be to "abuse" a password policy here. Currently users can configure the iteration count and default password hashing mechanism via policies:
> 
> ![image](https://user-images.githubusercontent.com/314690/156236839-cb990945-0e1c-4420-97f1-6045b432b44a.png) see: org.keycloak.policy.HashAlgorithmPasswordPolicyProviderFactory
> 
> Alternatively you could use lookup configuration values from realm attributes with your defaults as fallback.
> 
> @mposolda @stianst wdyt about making the number of generated backup codes, code lenght etc. configurable with a pw policy? `RecoveryCodePasswordPolicyProvider`? This would allow users to configure this.
> 
> Perhaps "CredentialPolicies" would have been a better level of abstraction for this.

Got something working, @thomasdarimont :

![Screen Shot 2022-03-01 at 2 28 31 PM](https://user-images.githubusercontent.com/91681638/156243928-9e5b4dd6-da26-4cad-b332-de066e52b359.png)

Corresponding admin screen:
![Screen Shot 2022-03-01 at 2 30 07 PM](https://user-images.githubusercontent.com/91681638/156244181-c2455b41-ce54-4daf-a949-261a835da0d6.png)

thomasdarimont commented:
> Got something working, @thomasdarimont :
Nice!
Now you could also add configuration properties for the number of recovery codes and recovery code hash algorithm :)

stianst commented on server-spi/src/main/java/org/keycloak/models/credential/PasswordCredentialModel.java:
```diff
@@ -44,8 +44,10 @@ public static PasswordCredentialModel createFromCredentialModel(CredentialModel
-            PasswordSecretData secretData = JsonSerialization.readValue(credentialModel.getSecretData(), PasswordSecretData.class);
-
+            PasswordSecretData secretData = null;
+            if (credentialModel.getSecretData() != null) {
```
How would it have been removed for security reasons? Why does this PR have any impact on this? Also, why would only the secret data be removed?

mposolda commented on themes/src/main/resources/theme/keycloak.v2/account/src/app/content/signingin-page/SigningInPage.tsx:
```diff
@@ -60,12 +60,27 @@ type CredType = string;
 type CredTypeMap = Map<CredType, CredentialContainer>;
 type CredContainerMap = Map<CredCategory, CredTypeMap>;
 
+interface CredData {
```
This interface `CredData` still contains some metadata specific to "recovery codes" credential. However it seems to me that this interface is not needed? Is it possible to remove this interface definition and also remove definition of `        const credData: CredData = JSON.parse(credential.credentialData!);` from line 271? It seems to me this won't break anything as `credData` defined on line 271 is not used anywhere?

stelewis-redhat replied:
OK. Removing. Good catch.

mposolda commented on core/src/main/java/org/keycloak/representations/idm/CredentialMetadataRepresentation.java:
```diff
@@ -0,0 +1,43 @@
+package org.keycloak.representations.idm;
+
+public class CredentialMetadataRepresentation {
```
Minor: IMO it will be slightly better to move this class into package `org.keycloak.representations.account` as it's main usage is for account REST API and account console right now

stelewis-redhat replied:
OK.

mposolda reviewed (changes requested):
@stelewis-redhat I did review of the things I proposed before, especially of the stuff related to credential metadata and `SigningInPage.tsx`, which was the biggest thing I proposed.

Things looks quite good to me, nice work! I've added few more comments related to that, but those are mostly minor things...

I agree with @stianst to not expand scope of this PR and postpone things like brute-force improvements as this is not specific to this authenticator, but other authenticator implementations we have in the codebase.

mposolda reviewed (changes requested):
Added few more inline comments

stelewis-redhat commented:
> > @stelewis-redhat let's not expand the scope of this PR to make everything configurable. I question that is needed in general, and it will also grow the scope of this PR a lot. So that should be a follow-up if not. For now let's focus on getting this ready to be merged.
> 
> Sorry, I didn't want to blow up the scope of this PR. However, I think being able to configure essential settings is an integral part of such a feature. But I agree, the feature is usable even without support for explicit configuration. However this should IMHO be added soon for completeness.
> 
> Some general guidance on feature completeness might help in the future? If we are striving to get something working initially, but it's not production grade then we should put it behind a preview flag. When something is "production grade" has to be defined.

OK I think I have all the other changes done. I need to do the preview flag. Looking into that now.

stelewis-redhat commented:
> @stianst Question: Should this be hidden behind PRofile/Feature flag and preview by default?

I'm looking at user profile as a reference to turn on a feature flag. It looks like there's changes to admin screens as well. Do I need to go that route or is there something smaller I can do, to do a feature flag? Is it all or nothing?

stelewis-redhat commented:
> 

@thomasdarimont I think I need some direction on the feature flag. I can enable/disable making the required action as default. I'm looking into removing recovery codes as an auth mechanism in the flow, but there doesn't seem to be a good hook there, as far as I can tell. Maybe there's a way to remove the required action from the available list.

mposolda commented:
> > @stelewis-redhat let's not expand the scope of this PR to make everything configurable. I question that is needed in general, and it will also grow the scope of this PR a lot. So that should be a follow-up if not. For now let's focus on getting this ready to be merged.
> 
> Sorry, I didn't want to blow up the scope of this PR. However, I think being able to configure essential settings is an integral part of such a feature. But I agree, the feature is usable even without support for explicit configuration. However this should IMHO be added soon for completeness.

@thomasdarimont Maybe we can see what is the feedback from community on this feature, and then based on that, we can decide if we need to improve and add more "configurability" and possibly something like "Policy configuration" . I personally hope that default settings can be sufficient for most of the cases, but maybe I am wrong here...

> Some general guidance on feature completeness might help in the future? If we are striving to get something working initially, but it's not production grade then we should put it behind a preview flag. When something is "production grade" has to be defined.

+1 for the "preview" flag

mposolda commented:
> > 
> 
> @thomasdarimont I think I need some direction on the feature flag. I can enable/disable making the required action as default. I'm looking into removing recovery codes as an auth mechanism in the flow, but there doesn't seem to be a good hook there, as far as I can tell. Maybe there's a way to remove the required action from the available list.
> 
> Specifically I need to know which places need to be driven by the feature flag. In theory, if we can keep the user from turning on the feature in admin, it'll never get into the account.v2 UI as a possible credential. Possible things driven by the flag:
> 
>     1. Keep user from adding required action.
> 
>     2. Keep user from adding auth to a flow.

@stelewis-redhat I think that it would be completely sufficient if you turn the provider factories (required action, authenticator, credential provider) to implement `EnvironmentDependentProviderFactory` interface and then implement the provider is "supported" just if feature is supported. You can take a look for example into `WebAuthnAuthenticatorFactory` for the inspiration.

Disabling providers will mean that administrator cannot register required action and cannot register authenticator and hence any "Recovery Code" stuff won't be visible at all for administrators or users. There are not any additional things needed in admin console or account console as AFAIK this feature doesn't introduce any "recovery codes" specific options into UI besides providers (For example WebAuthn was the bit more tricky as admin console has tabs "WebAuthn Policy" and "WebAuthn Passwordless Policy" and those should be hidden when WebAuthn feature is disabled. But "Recovery codes" doesn't have any such tab AFAIK, so you don't need to care about admin console or account console).

stelewis-redhat commented:
> > > 
> > 
> > 
> > @thomasdarimont I think I need some direction on the feature flag. I can enable/disable making the required action as default. I'm looking into removing recovery codes as an auth mechanism in the flow, but there doesn't seem to be a good hook there, as far as I can tell. Maybe there's a way to remove the required action from the available list.
> > Specifically I need to know which places need to be driven by the feature flag. In theory, if we can keep the user from turning on the feature in admin, it'll never get into the account.v2 UI as a possible credential. Possible things driven by the flag:
> > ```
> > 1. Keep user from adding required action.
> > 
> > 2. Keep user from adding auth to a flow.
> > ```
> 
> @stelewis-redhat I think that it would be completely sufficient if you turn the provider factories (required action, authenticator, credential provider) to implement `EnvironmentDependentProviderFactory` interface and then implement the provider is "supported" just if feature is supported. You can take a look for example into `WebAuthnAuthenticatorFactory` for the inspiration.
> 
> Disabling providers will mean that administrator cannot register required action and cannot register authenticator and hence any "Recovery Code" stuff won't be visible at all for administrators or users. There are not any additional things needed in admin console or account console as AFAIK this feature doesn't introduce any "recovery codes" specific options into UI besides providers (For example WebAuthn was the bit more tricky as admin console has tabs "WebAuthn Policy" and "WebAuthn Passwordless Policy" and those should be hidden when WebAuthn feature is disabled. But "Recovery codes" doesn't have any such tab AFAIK, so you don't need to care about admin console or account console).

Great thanks @mposolda . I have an implementation done already and am getting someone on my team to review my latest changes then I'll merge them in ASAP.

stelewis-redhat commented:
> > > > 
> > > 
> > > 
> > > @thomasdarimont I think I need some direction on the feature flag. I can enable/disable making the required action as default. I'm looking into removing recovery codes as an auth mechanism in the flow, but there doesn't seem to be a good hook there, as far as I can tell. Maybe there's a way to remove the required action from the available list.
> > > Specifically I need to know which places need to be driven by the feature flag. In theory, if we can keep the user from turning on the feature in admin, it'll never get into the account.v2 UI as a possible credential. Possible things driven by the flag:
> > > ```
> > > 1. Keep user from adding required action.
> > > 
> > > 2. Keep user from adding auth to a flow.
> > > ```
> > 
> > 
> > @stelewis-redhat I think that it would be completely sufficient if you turn the provider factories (required action, authenticator, credential provider) to implement `EnvironmentDependentProviderFactory` interface and then implement the provider is "supported" just if feature is supported. You can take a look for example into `WebAuthnAuthenticatorFactory` for the inspiration.
> > Disabling providers will mean that administrator cannot register required action and cannot register authenticator and hence any "Recovery Code" stuff won't be visible at all for administrators or users. There are not any additional things needed in admin console or account console as AFAIK this feature doesn't introduce any "recovery codes" specific options into UI besides providers (For example WebAuthn was the bit more tricky as admin console has tabs "WebAuthn Policy" and "WebAuthn Passwordless Policy" and those should be hidden when WebAuthn feature is disabled. But "Recovery codes" doesn't have any such tab AFAIK, so you don't need to care about admin console or account console).
> 
> Great thanks @mposolda . I have an implementation done already and am getting someone on my team to review my latest changes then I'll merge them in ASAP.

Everything I have outstanding, including feature flag, is merged into this PR. @stianst @mposolda @thomasdarimont

mposolda commented:
@stianst @thomasdarimont @stelewis-redhat  I missed some discussions related to password policy, but IMO password policies should be used to configure only things related to passwords. Using password policy for configure stuff related to "recovery codes" seems to me like a workaround and abusing password policies for something, which is not related to "password" at all.

I think it will be better to just hardcode the number to 4 to specify when warning for remaining codes should be displayed. And possibly expand this with "Recovery Code Policies" (or something like that) in some future release based on the feedback from the community.

So my vote is definitely to remove `RecoveryCodesWarningThresholdPasswordPolicyProviderFactory` and related stuff from this PR and just have the "4" as hardcoded number of remaining counts when warning is displayed.

stelewis-redhat commented:
> @stianst @thomasdarimont @stelewis-redhat I missed some discussions related to password policy, but IMO password policies should be used to configure only things related to passwords. Using password policy for configure stuff related to "recovery codes" seems to me like a workaround and abusing password policies for something, which is not related to "password" at all.
> 
> I think it will be better to just hardcode the number to 4 to specify when warning for remaining codes should be displayed. And possibly expand this with "Recovery Code Policies" (or something like that) in some future release based on the feedback from the community.
> 
> So my vote is definitely to remove `RecoveryCodesWarningThresholdPasswordPolicyProviderFactory` and related stuff from this PR and just have the "4" as hardcoded number of remaining counts when warning is displayed.

Removing.

stelewis-redhat commented:
> > @stianst @thomasdarimont @stelewis-redhat I missed some discussions related to password policy, but IMO password policies should be used to configure only things related to passwords. Using password policy for configure stuff related to "recovery codes" seems to me like a workaround and abusing password policies for something, which is not related to "password" at all.
> > I think it will be better to just hardcode the number to 4 to specify when warning for remaining codes should be displayed. And possibly expand this with "Recovery Code Policies" (or something like that) in some future release based on the feedback from the community.
> > So my vote is definitely to remove `RecoveryCodesWarningThresholdPasswordPolicyProviderFactory` and related stuff from this PR and just have the "4" as hardcoded number of remaining counts when warning is displayed.
> 
> Removing.

Removed @mposolda

stelewis-redhat commented:
> @stelewis-redhat The PR looks good to me! Thanks for all the updates to it!
> 
> I hope we can make this merged soon. I see the remaining steps are:
> 
> * Check the tests are ok (will wait until GH actions are finished in this PR. I hope there is not any issue here)

Is there anything I need to do on this? I fixed the tests that seemed to be directly related, but I'm unsure about the things that are failing. They seemed to be related to timing issues, last week. Things were timing out.

stelewis-redhat commented:
> * Documentation - the PR should be send to it to https://github.com/keycloak/keycloak-documentation . I think it will be good to update the sections:

Also, what do I need to do to send it to the keycloak-documentation project? Is there someone in particular I need to notify? Open an issue in that project?

mposolda commented:
> > * Documentation - the PR should be send to it to https://github.com/keycloak/keycloak-documentation . I think it will be good to update the sections:
> 
> Also, what do I need to do to send it to the keycloak-documentation project? Is there someone in particular I need to notify? Open an issue in that project?

Feel free to just send PR to the keycloak-documentation project and then comment here with the link to that PR. Also in the comment of the PR to keycloak-documentation, please add comment (or add it directly to the description of that PR) with the link to this "main codebase" PR.

stelewis-redhat commented:
> > > * Documentation - the PR should be send to it to https://github.com/keycloak/keycloak-documentation . I think it will be good to update the sections:
> > 
> > 
> > Also, what do I need to do to send it to the keycloak-documentation project? Is there someone in particular I need to notify? Open an issue in that project?
> 
> Feel free to just send PR to the keycloak-documentation project and then comment here with the link to that PR. Also in the comment of the PR to keycloak-documentation, please add comment (or add it directly to the description of that PR) with the link to this "main codebase" PR.

OK, so I'm updating the documentation as well?

mposolda commented:
> OK, so I'm updating the documentation as well?

Yes, it will be nice if you can do it.

stelewis-redhat commented:
> > OK, so I'm updating the documentation as well?
> 
> Yes, it will be nice if you can do it, so we can merge this PR together with the documentation.

No problem. Starting now.

stelewis-redhat commented:
> > OK, so I'm updating the documentation as well?
> 
> Yes, it will be nice if you can do it, so we can merge this PR together with the documentation.

https://github.com/keycloak/keycloak-documentation/pull/1429 created. I've linked this PR from that PR. @mposolda

mposolda commented:
@stelewis-redhat Sorry, but there is conflict as some other features were added into the "Profile" and "ProfileTest" in the meantime :/ Do you please have a chance to rebase this to latest "main" and fix the conflict?

stelewis-redhat commented:
> @stelewis-redhat Sorry, but there is conflict as some other features were added into the "Profile" and "ProfileTest" in the meantime :/ Do you please have a chance to rebase this to latest "main" and fix the conflict?

I'll look at it.

mposolda commented:
@stelewis-redhat Can I request one last update to the documentation inside this PR? See the comment https://github.com/keycloak/keycloak-documentation/pull/1429#issuecomment-1063992345 . If you can just update the section with preview features inside the file "features.adoc" and make sure that this feature for recovery codes is listed here?

stelewis-redhat commented:
> @stelewis-redhat Can I request one last update to the documentation inside this PR? See the comment [keycloak/keycloak-documentation#1429 (comment)](https://github.com/keycloak/keycloak-documentation/pull/1429#issuecomment-1063992345) . If you can just update the section with preview features inside the file "features.adoc" and make sure that this feature for recovery codes is listed here?

@mposolda it looks like features.adoc is driven by the java in Profile.java, so I think there's no more changes necessary?

mposolda commented:
@stelewis-redhat Good point :-) I am not very familiar with this stuff yet and didn't check it. Thanks again to you and everyone from your team for the patience with this task!

stelewis-redhat commented:
> @stelewis-redhat Good point :-) I am not very familiar with this stuff yet and didn't check it. Thanks again to you and everyone from your team for the patience with this task!

Thank you @mposolda @stianst and @thomasdarimont !

heddn commented:
I  know this is a bit off-topic, but when is v18.0.0 planned for release so this feature will be more generally available?
